text/2019-11-12-read-with-index.md

```diff
@@ -0,0 +1,60 @@
+# Follower Read With Applied Indices
+
+## Summary
+
+This RFC proposes a improvement about getting snapshot on followers, which is
+if the read request carries an applied index, the peer can get snapshot locally
+wihtout any commucation with its leader, if only it has applied to the given
+index. It's useful to reduce latency if the cluster is deployed in multi
+datacenters.
+
+## Motivation
+
+For clusters deployed in multi datacenters, the system latency could mainly
+depend on the network RTT between datacenters. For example, suppose PDs are
+deployed in Beijing, and TiKVs are deployed in Beijing and Xian (for high
+availability). If a client which is near to Xian wants to read a region, it
+needs to get a transaction timestamp from PDs at in Beijing, and then sends
+requests to TiKVs in Beijing or Xian. For the latter case, TiKVs in Xian will
+send read index requests to their leaders in Beijing internally, which still
+involves a RTT crossing datacenters. However, If the client can get timestamp
+and applied index for all target regions simultaneously, and then send read
+request to TiKVs with the timestamp and index, with the feature TiKVs can save
+one RTT, so the system latency can be reduced considerably.
+
+## Detailed design
+
+### Coprocessor Requests
+
+Add a field `applied_index` in `coprocessor.Request`:
+
+```protobuf
+message coprocessor.Request {
+    kvrpcpb.Context context = 1;
+    // omit other fields...
+
+    // After a region applys to `applied_index`, we can get a
+    // snapshot for the region even if the peer is follower.
+    uint64 applied_index = 5;
+}
```

Comment:
Why not put it into `kvrpcpb.Context` so other read requests can benefit from this as well.

Reply:
Good idea!